### DISCUSSION TITLE
<class 'AttributeError'> 'WSGIRequest' object has no attribute 'user'

### DISCUSSION THREAD

**Discussion by fvandonk** [netbox-community/netbox#13165] [Help Wanted!] [answered]
After doing what I assumed was a successful backup (3.5.4) and restore (3.5.5) now when I open the help on any page I get the following error:

Server Error
There was a problem with your request. Please contact an administrator.

The complete exception is provided below:

<class 'AttributeError'>

'WSGIRequest' object has no attribute 'user'

Python version: 3.9.16
NetBox version: 3.5.5

I checked the file structure and all the help files are there under: /opt/netbox/netbox/static/docs/

And the permissions for all the files are:
drwxr-xr-x.  root root 

Any idea why the help is now broken?

**Comment by candlerb** [comment]
Can you describe what you mean by "open the help on any page" - what exactly are you clicking on?

If you get a full backtrace, that may also shed some light. Either set DEBUG=True in configuration.py to get it in the web interface, or set ADMINS and EMAIL to have it E-mailed to you.

You could also try starting netbox as a [development server](https://docs.netbox.dev/en/stable/installation/3-netbox/#test-the-application).

**Reply by fvandonk** [reply]
The server is currently/still running as a development server.
As you recommended I set debug to True and now I get the following error when accessing any help page:

<SNIP>

_Page not found (404)
Directory indexes are not allowed here.
Request Method:	GET
Request URL:	http://172.17.136.235:8000/static/docs/models/dcim/rack/_

</SNIP>

I get to the help by click the help button located on each "+Add" page. 

![help-button](https://github.com/netbox-community/netbox/assets/43453987/682dd79c-dc23-4946-8c3a-4a02a41c32a6)

Console log with two help page access errors marked:
![2023-07-12_15-12-34](https://github.com/netbox-community/netbox/assets/43453987/90992d16-b51c-480e-8a96-f2baf270bdf5)

**Comment by candlerb** [comment]
> I get to the help by click the help button located on each "+Add" page.

Ah OK. Works for me with 3.5.6: e.g. from `/dcim/racks/add/` it redirects me to `/static/docs/models/dcim/rack/` (in a new tab) which renders fine.

On my system:

```
# ls /opt/netbox/netbox/static/docs/models/dcim/rack/
index.html
# 
```

What happens if you re-run `./upgrade.sh`? Are any errors shown?

**Reply by fvandonk** [reply]
I ran the upgrade and no errors were shown, but when I restarted the app the same error showed up.
 I wonder if this has something to do with running it on CentOS 9 stream. Can you tell me which OS and release you are running your instance on Netbox?

**Reply by candlerb** [reply]
Ubuntu 20.04, with apache2. But I don't think this is a SELinux issue - you would have got a permission error, not 404.

If you type `ls /opt/netbox/netbox/static/docs/models/dcim/rack/` what do you see?

If this directory contains an index.html then the issue is likely with your proxy (nginx/apache2).

**Reply by candlerb** [reply]
If index.html is not present, then it's likely a problem with the step that builds the documentation.

In that case, when you run `./upgrade.sh`, look carefully for the results at this step:

```
# Build the local documentation
COMMAND="mkdocs build"
echo "Building documentation ($COMMAND)..."
eval $COMMAND || exit 1
```

This is what I see:

```
...
Building documentation (mkdocs build)...
INFO     -  Cleaning site directory
INFO     -  Building documentation to directory: /opt/netbox/netbox/project-static/docs
INFO     -  The following pages exist in the docs directory, but are not included in the "nav" configuration:
              - index.md
INFO     -  Documentation built in 34.79 seconds
Collecting static files (python3 netbox/manage.py collectstatic --no-input)...

546 static files copied to '/opt/netbox/netbox/static', 24 unmodified.
Removing stale content types (python3 netbox/manage.py remove_stale_contenttypes --no-input)...
...
```

Also just to confirm, are you using the default root of `/opt/netbox`, not a custom location?

**Reply by fvandonk** [reply, marked as answer]
Hi Brian,
First off thank you for sticking with me on this. I really appreciate your help with this so far.

So here is where things stand. 
I first installed NetBox on Ubuntu as all steps provided are for Ubuntu and everything worked as expected, so that is good.
One thing to note is that if you stops after installation **step 3 NetBox** and run the server in development mode, the help pages do not work (even on the Ubuntu install). 

This is where you get the error that I talked about above:

<class 'AttributeError'>

'WSGIRequest' object has no attribute 'user'

Python version: 3.9.16
NetBox version: 3.5.5

**May I recommend to document this in the official documentation so that other people do not run into this same issue.**

The steps for either nginx and apache as provided **do not work** on CentOS 9 stream. 
I was however able to get it to work with a few changes and here are the steps in case you want to add them to the official documentation:

Installation step 5 HTTP Server
During the SSL command the location for the private key (_/etc/ssl/private_) does not exist on centos systems. You have to manually create it.
_sudo mkdir /etc/ssl/private_

Then if you install nginx and follow the commands the folder _/etc/nginx/sites-available/_ does not exist. 
If you continue you will find that the folder _/etc/nginx/sites-enabled/_ also does not exist .
Do not bother to create either because it still does not work because the nginx.conf file does not have any link to this folder so here is what I did to get it to work on nginx. This might not be the best way to do this, but it works :-)

I made a copy of the file _/etc/nginx/nginx.conf_ which is the main configuration file.
I copied all the content from the file _/opt/netbox/contrib/nginx.conf_ into memory.
I then edited the file _/etc/nginx/nginx.conf_ 
There are multiple 'sections' in this file like *events* and *http* under *http* is a subsection named *server* 
Completely remove the section:
_server {
...
}_

You can leave the commented text for _Settings for a TLS enabled server_ as it is not used.

Then paste from memory the text from the _/opt/netbox/contrib/nginx.conf_ file which has two server sections:
server {
...
}

One for port 443 and one for port 80 to redirect to port 433.
So you are pasting both server sections into the location where you removed the default server section.

Then update this section:
    **# CHANGE THIS TO YOUR SERVER'S NAME
    server_name netbox.example.com;**
Save and close the file 

Restart the service
_sudo systemctl restart nginx_

Test and verify that the site is up and the help is working.

After making these changes I have a working NetBox 3.5.6 on CentOS 9 stream                                       

Thanks again,
Fred
